Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift

```diff
@@ -0,0 +1,126 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+extension SymbolGraph {
+    /// Create "overload group" symbols based on name and kind collisions.
+    ///
+    /// For this method, an "overload" is a symbol whose ``Symbol/pathComponents`` and
+    /// ``Symbol/kind-swift.property`` are the same as another symbol in the same symbol graph.
+    /// Such symbols are usually found in languages that allow for function overloading based on
+    /// parameter or return types.
+    ///
+    /// When this method is called, it first looks for any symbols with an overloadable symbol kind
+    /// (see ``Symbol/KindIdentifier/isOverloadableKind``) which collide on both kind and path. It
+    /// then sorts these colliding symbols in one of two ways:
+    ///
+    /// 1. If all the colliding symbols have ``Symbol/DeclarationFragments-swift.struct``, these
+    ///    declarations are condensed into strings by their
+    ///    ``Symbol/DeclarationFragments-swift.struct/Fragment/spelling``, which are then sorted.
+    /// 2. Otherwise, the symbols are sorted by their unique identifier.
+    ///
+    /// The symbol that appears first in this sorting is then cloned to create an "overload group".
+    /// This symbol will have a unique identifier based on the original symbol, but suffixed with
+    /// ``Symbol/overloadGroupIdentifierSuffix``. New ``Relationship/Kind/overloadOf``
+    /// relationships are created between the colliding symbols and the new overload group symbol.
+    /// In addition, any existing relationships the original symbol had are also cloned for the
+    /// overload group.
+    public mutating func createOverloadGroupSymbols() {
+        struct OverloadKey: Hashable {
+            let path: [String]
+            let kind: SymbolGraph.Symbol.KindIdentifier
+        }
+
+        var symbolsByPath = [OverloadKey: [SymbolGraph.Symbol]]()
+
+        for symbol in self.symbols.values {
+            if symbol.kind.identifier.isOverloadableKind {
+                symbolsByPath[.init(path: symbol.pathComponents, kind: symbol.kind.identifier), default: []].append(symbol)
+            }
+        }
```

Comment:
Dictionary has an initializer that groups elements. That way `symbolsByPath` can remain immutable throughout the rest of this implementation:
```suggestion
        let symbolsByPath = [OverloadKey: [SymbolGraph.Symbol]](
            grouping: symbols.values.filter(\.kind.identifier.isOverloadableKind),
            by: { .init(path: $0.pathComponents, kind: $0.kind.identifier) }
        )
```